KIPs/kip-dreaft_sdk_common_architecture.md

```diff
@@ -0,0 +1,928 @@
+
+The `ABI` layer provides the functions to encode and decode parameters based on ABI.
+
+The `Transaction` layer contains classes of Klaytn transaction types, `TransactionDecoder`, and `TransactionHasher`.
+
+The `TransactionDecoder` class decodes the RLP-encoded transaction string. The `TransactionHasher` class calculates the hash of the transaction.
+
+The `Wallet` layer contains keyring classes (`SingleKeyring`, `MultipleKeyring`, and `RoleBasedKeyring`) and `KeyringContainer` class that contains multiple keyrings. The `KeyringContainer` class acts as an "in-memory wallet" that stores the Keyring instances.
+
+The `Account` layer contains an `Account` class that stores information needed when updating the [AccountKey] of the account in the Klaytn.
+
+Also, the `RPC` layer includes the `Klay` class, which is responsible for rpc calls of the klay namespace, and the `Net` class, which is responsible for rpc calls of the net namespace.
+
+Finally, the `Utils` layer contains the `Utils` class that provides utility functions.
+
+From the next chapter, each layer is described in detail.
+
+### Account Layer Class Diagram
+
+The `Account` layer provides functionality related to updating the [AccountKey] of the Klaytn account.
+
+![account](https://user-images.githubusercontent.com/32922423/86084568-2da96580-bad8-11ea-9c46-e74ae6d8ad7a.png)
+
+`Account` is a class that contains information needed to update the [AccountKey] of the account in the Klaytn. It has `address` and `accountKey` as member variables. The `accountKey` can be an instance of (`AccountKeyLegacy`, `AccountKeyPublic`, `AccountKeyFail`, `AccountKeyWeightedMultiSig` or `AccountKeyRoleBased`) depending on the key.
+
+An `Account` instance is injected into the `account` field of the AccountUpdate transaction types (`AccountUpdate`, `FeeDelegatedAccountUpdate` or `FeeDelegatedAccountUpdateWithRatio`). When an AccountUpdate transaction is successfully processed in Klaytn, the account key of the [Klaytn account] is updated with the `accountKey` defined in the `Account` instance.
+
+Interface for AccountKey is defined as `IAccountKey`, and account key classes (`AccountKeyLegacy`, `AccountKeyPublic`, `AccountKeyFail`, `AccountKeyWeightedMultiSig` and `AccountKeyRoleBased`) implements IAccountKey.
+
+`WeightedMultiSigOptions` is provided for `AccountKeyWeightedMultiSig` where the threshold and the weight of each key are defined.
+
+The `AccountKeyDecoder` class decodes the RLP-encoded string using the decode function implemented in each AccountKey class.
+
+#### IAccountKey
+
+| Method | Description |
+| ----------- | ----------- |
+| getRLPEncoding(): String | Returns a RLP-encoded string of AccountKey. AccountKey classes implement `IAccountKey`, and this function must be implemented. |
+
+#### AccountKeyLegacy
+
+| Method | Description |
+| ----------- | ----------- |
+| decode(rlpEncodedKey: String): AccountKeyLegacy | Decodes a RLP-encoded string of AccountKeyLegacy. |
```

Comment:
```suggestion
| decode(rlpEncodedKey: String): AccountKeyLegacy | Decodes an RLP-encoded string of AccountKeyLegacy and returns a `AccountKeyLegacy` object. |
```

Reply:
@jimni1222 What happens if the decode process fails? it throws? or returns a null object?

Reply:
It will throw an error!

Reply:
@jimni1222 Then, please insert the sentence below for the relevant cases:

It throws an exception if the decoding is failed.